src/authentication/MobileRouter.jsx

```diff
@@ -1,39 +1,131 @@
+        cozy_url,
+        this.props.client
+      )
+
+      const { onboarding_secret, onboarding_state } = clientInfo
+      if (
+        !checkExchangedInformations(
+          localSecret,
+          onboarding_secret,
+          localState,
+          onboarding_state
+        )
+      )
+        throw new Error('exchanged informations are not good')
 
-const MobileRouter = ({
-  history,
-  appRoutes,
-  isAuthenticated,
-  isRevoked,
-  onAuthenticated,
-  onLogout,
-  appIcon
-}) => {
-  if (!isAuthenticated) {
-    return (
-      <Authentication
-        router={history}
-        onComplete={onAuthenticated}
-        onException={logException}
-        appIcon={appIcon}
-      />
-    )
-  } else if (isRevoked) {
-    return (
-      <Revoked
-        router={history}
-        onLogBackIn={onAuthenticated}
-        onLogout={onLogout}
-      />
-    )
-  } else {
-    return <Router history={history}>{appRoutes}</Router>
+      const getTokenRequest = await getAccessToken(
+        clientInfo,
+        cozy_url,
+        code,
+        this.props.client
+      )
+      const token = await getTokenRequest.json()
+      if (getTokenRequest.status !== 200) {
+        throw new Error('token.error')
+      }
+
+      await this.props.onAuthenticated({
+        url: `https://${cozy_url}`,
+        token,
+        clientInfo,
+        router: history
+      })
+    } catch (error) {
+      this.props.onLogout()
+    } finally {
+      clearState()
+      clearSecret()
+    }
+  }
+  render() {
+    const {
+      history,
+      appRoutes,
+      isAuthenticated,
+      isRevoked,
+      onAuthenticated,
+      onLogout,
+      appIcon,
+      onboarding,
+      onboardingInformations
+    } = this.props
+
+    if (!isAuthenticated) {
+      if (checkIfOnboardingLogin(onboardingInformations)) {
+        /* We need to hide() the ViewController since the ViewController is still active 
+        when the application cames from background (specialy on iOS)
+        */
+        if (window.SafariViewController) window.SafariViewController.hide()
+        const { code, state, cozy_url } = onboardingInformations
+        this.doOnboardingLogin(state, code, cozy_url, history)
+        //we return null since the previous method is async
+        return null
+      } else {
+        return (
+          <Authentication
+            router={history}
+            onComplete={onAuthenticated}
+            onException={logException}
+            appIcon={appIcon}
+            onboarding={onboarding}
+          />
+        )
+      }
+    } else if (isRevoked) {
+      return (
+        <Revoked
+          router={history}
+          onLogBackIn={onAuthenticated}
+          onLogout={onLogout}
+        />
+      )
+    } else {
+      return <Router history={history}>{appRoutes}</Router>
+    }
   }
 }
 
+MobileRouter.propTypes = {
+  onboarding: onboardingPropTypes.isRequired,
+  onboardingInformations: onboardingInformationsPropTypes.isRequired,
```

Comment:
Now that most of the review is behind us, I have some more nits — I think `onboarding ` and `onboardingInformations ` are not so great names.
`onboarding` should be something like `oauthInfos` I think, and `onboardingInformations ` maybe `autloginInformations`. But none of them really do much onboarding.